fish-rust/src/io.rs

```diff
@@ -343,10 +344,13 @@ impl IoBufferfill {
         // Our buffer will read from the read end of the pipe. This end must be non-blocking. This is
         // because our fillthread needs to poll to decide if it should shut down, and also accept input
         // from direct buffer transfers.
-        if ffi::make_fd_nonblocking(autocxx::c_int(pipes.read.fd())).0 != 0 {
-            FLOG!(warning, PIPE_ERROR);
-            perror("fcntl");
-            return None;
+        match make_fd_nonblocking(&pipes.read.fd()) {
+            Ok(_) => (),
+            Err(e) => {
+                FLOG!(warning, PIPE_ERROR);
+                perror_io("fcntl", &e);
```

Comment:
Going forward we should move the information about fcntl into the callee. Perhaps by making our custom error type. Something like

```rust
pub struct LibcError { // or SyscallError, or simply Error
	causer: &'static str,
	error: io::Error,
}
```

and define `std::fmt::Display` to print it as usual.

Here we'd get a `LibcError{causer: "fcntl", error: io::Error::last_os_error()}`

Reply:
Why not just use nix’s (some are even safe!) wrappers? https://docs.rs/nix/latest/nix/fcntl/index.html

Reply:
I don't think the error returned by Nix's fcntl includes an indication that it was fcntl that failed but yes we should cut back on raw libc calls. I mostly use raw libc calls because that's the easiest for porting. It probably wouldn't be a mistake to do so here as well, to make sure that porting callers is trivial.

Reply:
Agree with all of this but I think it doesn't have to be part of this PR

Reply:
OK, I've left it as is for now. Possibly using straight `perror` is fine, but because there's a `FLOG` first it's theoretically not correct.